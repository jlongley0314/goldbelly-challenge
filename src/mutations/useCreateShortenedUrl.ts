import { useMutation } from "react-query";

export type UrlShortenerFormData = {
  url: string;
  short_url: string;
  slug?: string;
};

export function useCreateShortenedUrl(accessToken: string) {
  return useMutation(async (formData: UrlShortenerFormData) => {
    let headers = new Headers();
    headers.append("GB-Access-Token", accessToken);
    headers.append("Content-Type", "application/json");
    const body = JSON.stringify(formData);
    const requestOptions = {
      method: "POST",
      headers: headers,
      body: body,
    };
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/links`,
      requestOptions
    );

    if (!response.ok) {
      const error = await response.json();
      if (error.errors && Object.keys(error.errors)[0]) {
        const errorKey = Object.keys(error.errors)[0];
        const errorValue = Object.values(error.errors)[0];
        const newError = new Error();
        newError.message = `${errorKey} ${errorValue}`;
        throw newError;
      } else {
        throw new Error("Something went wrong.");
      }
    }

    return response;
  });
}
