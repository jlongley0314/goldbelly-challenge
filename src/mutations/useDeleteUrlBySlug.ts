import { useMutation } from "react-query";

export function useDeleteUrlBySlug(accessToken: string) {
  return useMutation(async (slug: string) => {
    let headers = new Headers();
    headers.append("GB-Access-Token", accessToken);
    headers.append("Content-Type", "application/json");
    const requestOptions = {
      method: "DELETE",
      headers: headers,
    };
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/links/${slug}`,
      requestOptions
    );

    if (!response.ok) {
      const error = await response.json();
      throw error.error;
    }

    return response;
  });
}
