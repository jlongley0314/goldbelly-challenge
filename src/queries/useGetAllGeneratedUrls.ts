import { useQuery } from "react-query";

async function getAllGeneratedUrls(accessToken: string) {
  let myHeaders = new Headers();
  myHeaders.append("GB-Access-Token", accessToken);
  const response = await fetch(`${process.env.REACT_APP_API_URL}/links`, {
    method: "GET",
    headers: myHeaders,
  });
  if (!response.ok) {
    throw new Error("Error fetching generated urls");
  }
  return await response.json();
}

export function useGetAllGeneratedUrls(accessToken: string) {
  return useQuery(
    "getAllGeneratedUrls",
    () => getAllGeneratedUrls(accessToken),
    {
      enabled: !!accessToken && accessToken !== "",
    }
  );
}
