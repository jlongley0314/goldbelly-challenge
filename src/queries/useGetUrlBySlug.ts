import { useQuery } from "react-query";

async function getUrlBySlug(accessToken: string, slug: string) {
  let myHeaders = new Headers();
  myHeaders.append("GB-Access-Token", accessToken);
  const response = await fetch(
    `${process.env.REACT_APP_API_URL}/links/${slug}`,
    {
      method: "GET",
      headers: myHeaders,
    }
  );
  if (!response.ok) {
    throw new Error(`Error fetching url with slug: ${slug}`);
  }
  return await response.json();
}

export function useGetUrlBySlug(accessToken: string, slug: string) {
  return useQuery(
    `getUrlBySlug:${slug}`,
    () => getUrlBySlug(accessToken, slug),
    {
      enabled: !!accessToken && accessToken !== "" && !!slug && slug !== "",
    }
  );
}
