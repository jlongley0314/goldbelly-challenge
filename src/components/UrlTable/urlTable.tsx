import "./urlTable.css";
import Card from "react-bootstrap/Card";
import Table from "react-bootstrap/Table";
import Button from "react-bootstrap/Button";
import { useGetAllGeneratedUrls } from "../../queries/useGetAllGeneratedUrls";
import { useDeleteUrlBySlug } from "../../mutations/useDeleteUrlBySlug";
import { useEffect } from "react";
import { useQueryClient } from "react-query";

type UrlShortenerDataResponse = {
  url: string;
  short_url: string;
  slug: string;
};

export function UrlTable() {
  const queryClient = useQueryClient();
  const getAllGeneratedUrlsQuery = useGetAllGeneratedUrls(
    process.env.REACT_APP_API_KEY ?? ""
  );

  const { mutate: deleteUrl, isSuccess: deleteSuccess } = useDeleteUrlBySlug(
    process.env.REACT_APP_API_KEY ?? ""
  );

  useEffect(() => {
    if (deleteSuccess) {
      queryClient.invalidateQueries("getAllGeneratedUrls");
    }
  }, [deleteSuccess, queryClient]);

  return (
    <Card>
      <Card.Title className="UrlTableHeader">Your URLs</Card.Title>
      <Card.Body>
        <Table striped bordered hover>
          <thead>
            <tr>
              <th>Long Url</th>
              <th>Short Url</th>
              <th>Slug</th>
              <th></th>
            </tr>
          </thead>
          <tbody>
            {getAllGeneratedUrlsQuery.isSuccess &&
              getAllGeneratedUrlsQuery.data.map(
                (data: UrlShortenerDataResponse) => {
                  return (
                    <tr key={data.short_url}>
                      <td>{data.url}</td>
                      <td>{data.short_url}</td>
                      <td>{data.slug}</td>
                      <td>
                        <Button
                          variant="secondary"
                          onClick={() => deleteUrl(data.slug)}
                        >
                          Delete
                        </Button>
                      </td>
                    </tr>
                  );
                }
              )}
          </tbody>
        </Table>
      </Card.Body>
    </Card>
  );
}
