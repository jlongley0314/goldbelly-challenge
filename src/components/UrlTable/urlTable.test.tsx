import {
  fireEvent,
  render,
  RenderResult,
  screen,
} from "@testing-library/react";
import { useQueryClient } from "react-query";
jest.mock("react-query");
import { useDeleteUrlBySlug } from "../../mutations/useDeleteUrlBySlug";
jest.mock("../../mutations/useDeleteUrlBySlug");
import { useGetAllGeneratedUrls } from "../../queries/useGetAllGeneratedUrls";
jest.mock("../../queries/useGetAllGeneratedUrls");
import { UrlTable } from "./urlTable";

const mockUrls = [
  {
    url: "fdsafs",
    slug: "shorter",
    short_url: "http://bely.me/shorter",
  },
  {
    url: "asdfasdf",
    slug: "s",
    short_url: "http://bely.me/s",
  },
  {
    url: "asdfdsafdsa",
    slug: "tewrewfsdaf",
    short_url: "http://bely.me/tewrewfsdaf",
  },
  {
    url: "fdsafsdafdas",
    slug: "dfasdfsdafsdafsdafd",
    short_url: "http://bely.me/dfasdfsdafsdafsdafd",
  },
  {
    url: "fdsacdsacdasc",
    slug: "cdsacdsacd",
    short_url: "http://bely.me/cdsacdsacd",
  },
  {
    url: "asdfdsfasdcdasc",
    slug: "dcasddfsdafdsaf",
    short_url: "http://bely.me/dcasddfsdafdsaf",
  },
];

describe("URLTable", () => {
  let component: RenderResult;
  let mutate = jest.fn();
  let invalidateQueries = jest.fn();

  beforeEach(() => {
    (useDeleteUrlBySlug as jest.Mock).mockReturnValue({
      isLoading: false,
      isError: false,
      error: undefined,
      isSuccess: true,
      mutate: mutate,
    });
    (useQueryClient as jest.Mock).mockReturnValue({ invalidateQueries });
    (useGetAllGeneratedUrls as jest.Mock).mockReturnValue({
      data: mockUrls,
      isSuccess: true,
    });
    component = render(<UrlTable />);
  });

  it("should render component", () => {
    expect(component).toMatchSnapshot();
  });

  it("should display the url values in table", () => {
    mockUrls.forEach((url) => {
      expect(screen.getByText(url.url)).toBeInTheDocument;
      expect(screen.getByText(url.short_url)).toBeInTheDocument;
      expect(screen.getByText(url.slug)).toBeInTheDocument;
    });
  });

  it("should call delete mutation with correct slug when delete button is pressed", () => {
    const url = mockUrls[0];
    fireEvent.click(screen.getAllByText("Delete")[0]);
    expect(mutate).toHaveBeenCalledWith(url.slug);
  });
});
