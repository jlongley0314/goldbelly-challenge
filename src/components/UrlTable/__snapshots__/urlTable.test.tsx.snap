// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`URLTable should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="UrlTableHeader card-title h5"
        >
          Your URLs
        </div>
        <div
          class="card-body"
        >
          <table
            class="table table-striped table-bordered table-hover"
          >
            <thead>
              <tr>
                <th>
                  Long Url
                </th>
                <th>
                  Short Url
                </th>
                <th>
                  Slug
                </th>
                <th />
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  fdsafs
                </td>
                <td>
                  http://bely.me/shorter
                </td>
                <td>
                  shorter
                </td>
                <td>
                  <button
                    class="btn btn-secondary"
                    type="button"
                  >
                    Delete
                  </button>
                </td>
              </tr>
              <tr>
                <td>
                  asdfasdf
                </td>
                <td>
                  http://bely.me/s
                </td>
                <td>
                  s
                </td>
                <td>
                  <button
                    class="btn btn-secondary"
                    type="button"
                  >
                    Delete
                  </button>
                </td>
              </tr>
              <tr>
                <td>
                  asdfdsafdsa
                </td>
                <td>
                  http://bely.me/tewrewfsdaf
                </td>
                <td>
                  tewrewfsdaf
                </td>
                <td>
                  <button
                    class="btn btn-secondary"
                    type="button"
                  >
                    Delete
                  </button>
                </td>
              </tr>
              <tr>
                <td>
                  fdsafsdafdas
                </td>
                <td>
                  http://bely.me/dfasdfsdafsdafsdafd
                </td>
                <td>
                  dfasdfsdafsdafsdafd
                </td>
                <td>
                  <button
                    class="btn btn-secondary"
                    type="button"
                  >
                    Delete
                  </button>
                </td>
              </tr>
              <tr>
                <td>
                  fdsacdsacdasc
                </td>
                <td>
                  http://bely.me/cdsacdsacd
                </td>
                <td>
                  cdsacdsacd
                </td>
                <td>
                  <button
                    class="btn btn-secondary"
                    type="button"
                  >
                    Delete
                  </button>
                </td>
              </tr>
              <tr>
                <td>
                  asdfdsfasdcdasc
                </td>
                <td>
                  http://bely.me/dcasddfsdafdsaf
                </td>
                <td>
                  dcasddfsdafdsaf
                </td>
                <td>
                  <button
                    class="btn btn-secondary"
                    type="button"
                  >
                    Delete
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
    >
      <div
        class="UrlTableHeader card-title h5"
      >
        Your URLs
      </div>
      <div
        class="card-body"
      >
        <table
          class="table table-striped table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                Long Url
              </th>
              <th>
                Short Url
              </th>
              <th>
                Slug
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                fdsafs
              </td>
              <td>
                http://bely.me/shorter
              </td>
              <td>
                shorter
              </td>
              <td>
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Delete
                </button>
              </td>
            </tr>
            <tr>
              <td>
                asdfasdf
              </td>
              <td>
                http://bely.me/s
              </td>
              <td>
                s
              </td>
              <td>
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Delete
                </button>
              </td>
            </tr>
            <tr>
              <td>
                asdfdsafdsa
              </td>
              <td>
                http://bely.me/tewrewfsdaf
              </td>
              <td>
                tewrewfsdaf
              </td>
              <td>
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Delete
                </button>
              </td>
            </tr>
            <tr>
              <td>
                fdsafsdafdas
              </td>
              <td>
                http://bely.me/dfasdfsdafsdafsdafd
              </td>
              <td>
                dfasdfsdafsdafsdafd
              </td>
              <td>
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Delete
                </button>
              </td>
            </tr>
            <tr>
              <td>
                fdsacdsacdasc
              </td>
              <td>
                http://bely.me/cdsacdsacd
              </td>
              <td>
                cdsacdsacd
              </td>
              <td>
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Delete
                </button>
              </td>
            </tr>
            <tr>
              <td>
                asdfdsfasdcdasc
              </td>
              <td>
                http://bely.me/dcasddfsdafdsaf
              </td>
              <td>
                dcasddfsdafdsaf
              </td>
              <td>
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Delete
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
